### PR TITLE
Ignore words for various rules

### PR DESCRIPTION
**Pull request by tunetheweb**
### Brief summary of the change made

As raised on Slack 0.9.2 introduced a lot of keywords to Snowflake, including a lot of single word short alias's for dates (e.g. `d` for `day`).

This PR introduces the concept of `ignore_words` to allow the rules that check identifiers to be ignored for certain "allowed words". This is a good feature for feature to have anyway as allows users to handle bugs without waiting for the next release or turning off a rule entirely.

I've also excluded single letter chars from L026 and did some clean up in there for performance.

Adds an optional `ignore_words` parameter to the following rules:
- L010 - Inconsistent capitalisation of keywords.
- L014 - Inconsistent capitalisation of unquoted identifiers.
- L029 - Keywords should not be used as identifiers (rule inherits this from L010).
- L030 - Inconsistent capitalisation of function names (rule inherits this from L010).
- L040 - Inconsistent capitalisation of boolean/null literal (rule inherits this from L010).

I've also made some further changes to L029:
- Exit early if not an identifier
- Exit if 1 character identifier (sometimes these are keywords - like `d` in Snowflake, but people expect to be able to use them)
- Cache cleaned up `ignore_words` list (similar to how L010 caches it's config).

### Are there any other side effects of this change that we should be aware of?
Shouldn't be

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: test/fixtures/rules/std_rule_cases/L040.yml [hunk 1]
```diff
@@ -3,3 +3,10 @@ rule: L040
 test_fail_inconsistent_boolean_capitalisation:
   fail_str: SeLeCt true, FALSE, NULL
   fix_str: SeLeCt true, false, null
+
+test_pass_ignore_word:
+  pass_str: SELECT true, FALSE, NULL
+  configs:
+    rules:
+      L028:
```
Comment (tunetheweb) [inline test/fixtures/rules/std_rule_cases/L040.yml:11]:
```suggestion
      L040:
```

File: test/fixtures/rules/std_rule_cases/L030.yml [hunk 2]
```diff
@@ -47,3 +47,10 @@ test_fail_fully_qualified_function_mixed_functions:
 test_fail_fully_qualified_function_mixed_case:
   fail_str: SELECT project1.FoO(value1) AS value2
   fix_str: SELECT project1.FOO(value1) AS value2
+
+test_pass_ignore_word:
+  pass_str: SELECT MAX(id), min(id) FROM TABLE1
+  configs:
+    rules:
+      L028:
```
Comment (tunetheweb) [inline test/fixtures/rules/std_rule_cases/L030.yml:55]:
```suggestion
      L030:
```

File: src/sqlfluff/core/rules/config_info.py [hunk 3]
```diff
@@ -81,6 +81,9 @@
             " be ignored when linting line lengths?"
         ),
     },
+    "ignore_words": {
+        "definition": ("List or words to ignore from rule"),
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/config_info.py:85]:
```suggestion
        "definition": ("List of words to ignore from rule"),
```

File: src/sqlfluff/rules/L010.py [hunk 4]
```diff
@@ -191,6 +200,10 @@ def _get_fix(self, segment, fixed_raw):
         """
         return LintFix.replace(segment, [segment.edit(fixed_raw)])
 
+    @staticmethod
+    def _split_comma_separated_string(raw_str: str) -> List[str]:
```
Comment (barrywhart) [inline src/sqlfluff/rules/L010.py:204]: This function is pretty generic, and we're repeating it several times in this PR. Should we consider adding it to `BaseRule` instead?

File: test/fixtures/rules/std_rule_cases/L029.yml [hunk 5]
```diff
@@ -114,3 +114,25 @@ test_fail_keyword_as_quoted_identifier_column_alias_config:
...
+    rules:
+      L029:
+        ignore_word: date
+
+
+test_pass_one_character_identifier:
+  pass_str: SELECT d.col1 FROM table1 d
+  configs:
+    core:
+      dialect: Snowflake
```
Comment (barrywhart) [inline test/fixtures/rules/std_rule_cases/L029.yml:138]: Nitpick: Did you mean to capitalize the "S" in "snowflake"? It's the first time I've seen capital letters when referring to dialects.

**Review by barrywhart — CHANGES_REQUESTED** [review]

File: src/sqlfluff/rules/L010.py (hunk 4 above)
Comment (tunetheweb) [reply src/sqlfluff/rules/L010.py:204]: Yeah was wondering where best place to put it. That's sounds good to me. Think it's used elsewhere too so can clean those up at same time.

File: test/fixtures/rules/std_rule_cases/L029.yml (hunk 5 above)
Comment (tunetheweb) [reply test/fixtures/rules/std_rule_cases/L029.yml:138]: No - that's a good spot! It doesn't work with that, hence tests failing.

File: src/sqlfluff/core/rules/base.py [hunk 6]
```diff
@@ -773,6 +773,10 @@ def discard_unsafe_fixes(
                 lint_result.fixes = []
                 return
 
+    @staticmethod
+    def _split_comma_separated_string(raw_str: str) -> List[str]:
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/base.py:777]: Nitpick: If we remove the `_` prefix, I think this function would show up in the "autodoc" docs for the `BaseRule` class, which could be helpful for other people working on rules.

**Review by barrywhart — APPROVED** [review]

File: src/sqlfluff/core/rules/base.py (hunk 6 above)
Comment (tunetheweb) [reply src/sqlfluff/core/rules/base.py:777]: Will do. Also wonder if we should add BaseRule to https://docs.sqlfluff.com/en/stable/developingrules.html

Will open a separate PR for that.

**Pull request merged**